### PR TITLE
Subset Sum Problem

### PR DESCRIPTION
**Pull request by Kroco23**
Implemented brute force solution and two optimized algorithms for subset sum problem

### CODE REVIEW

File: subset/SSet.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+bool SSet::checkSumOfNSlow(DataType expected, size_t N) {
+  std::vector<DataType> sub_set(N);
+  return checkSumOfNRecursive(sub_set, expected, N, 0, 0);
+}
+
+/* data   ---> Temporary array to store current combination
+   N      ---> Size of a combination to be checked
+   index  ---> Current index in data
+   i      ---> index of current element in set_     */
+bool SSet::checkSumOfNRecursive(std::vector<DataType> data, DataType expected,
```
Comment (denislos) [inline subset/SSet.cpp:84]: It is better to pass std::vector as reference here to avoid unnecessary copying

File: subset/SSet.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+bool SSet::checkSumOptimizedRecursive(int i, DataType expected) {
+  if (i == 0) return set_[i] == expected;
+  if ((expected > positive_sum) || (expected < negative_sum)) return false;
+  return checkSumOptimizedRecursive(i - 1, expected) || (set_[i] == expected) ||
+         checkSumOptimizedRecursive(i - 1, expected - set_[i]);
+}
+
+std::vector<SSet::DataType> SSet::subsetSums(std::vector<DataType>* set_p) {
```
Comment (alex19999) [inline subset/SSet.cpp:23]: Quite strange function. Take pointer to an vector and copy vector itself. Then use this copy for reading elements. This is redundant

File: subset/SSet.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+  std::sort(firstlist.begin(), firstlist.end());
+  std::sort(secondlist.begin(), secondlist.end());
+
+  for (firstit = firstlist.begin(); firstit != firstlist.end(); firstit++) {
+    for (secondit = secondlist.begin(); secondit != secondlist.end(); secondit++) {
+      if ((*firstit == 0) || (*secondit == 0)) return true;
+      int sum = *firstit + *secondit;
+      if (sum == expected) return true;
+      else if (sum > expected) break;
```
Comment (alex19999) [inline subset/SSet.cpp:59]: "else" keyword is redundant because when if-statement is true it calls return

File: subset/SSet.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+  }
+  return false;
+}
+
+bool SSet::checkSumSlow(DataType expected) {
+  for (size_t i = 1; i <= set_.size(); i++) {
+    if (checkSumOfNSlow(expected, i)) {
```
Comment (alex19999) [inline subset/SSet.cpp:67]: Codestyle: use or don't use {} for one line statement everywhere

File: subset/SSet.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  using DataType = int;
+
+  void addElement(DataType elem) { set_.push_back(elem); }
+
+  bool checkZeroSumSlow() { return checkSumSlow(0); }
+
+  bool checkZeroSumOptimized() { return checkSumOptimized(0); }
+
+  bool checkZeroSumOptimizedHS() { return checkSumOptimizedHS(0); }
+std::vector<DataType> subsetSums(std::vector<DataType>* set);
```
Comment (alex19999) [inline subset/SSet.h:21]: Fix indent

File: subset/SSet.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    for (auto el : set_) {
+      std::cout << el << " ";
+    }
+    std::cout << "\n";
+  }
+
+ private:
+  DataType sumSet(std::vector<DataType> set) {
+    DataType sum = 0;
+    for (auto el : set) {
```
Comment (alex19999) [inline subset/SSet.h:55]: See std::accumulate

File: subset/SSet.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  void print() const {
+    std::cout << "SSet: ";
+    for (auto el : set_) {
+      std::cout << el << " ";
+    }
+    std::cout << "\n";
+  }
+
+ private:
+  DataType sumSet(std::vector<DataType> set) {
```
Comment (denislos) [inline subset/SSet.h:53]: Same thing with copying std::vector

File: subset/bench_subset_optHS.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+BENCHMARK(BM_subsetOPTHSFromFile10T);
+BENCHMARK(BM_subsetOPTHSFromFile10F);
+BENCHMARK(BM_subsetOPTHSFromFile15T);
+BENCHMARK(BM_subsetOPTHSFromFile15F);
+BENCHMARK(BM_subsetOPTHSFromFile20T);
+BENCHMARK(BM_subsetOPTHSFromFile20F);
+BENCHMARK(BM_subsetOPTHSFromFile25T);
+BENCHMARK(BM_subsetOPTHSFromFile25F);
+BENCHMARK(BM_subsetOPTHSFromFile30T);
+BENCHMARK(BM_subsetOPTHSFromFile30F);
```
Comment (denislos) [inline subset/bench_subset_optHS.cpp:86]: It is a myriad of merely the same functions. Maybe it is possible to create a single function and pass trace file into as a parameter

File: subset/SSet.cpp [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      prevPowOf2 *= 2;
+    }
+    sums[i] = set[effectiveBits - 1] + sums[i - prevPowOf2];
+  }
+  sums.erase(sums.begin());
+  return sums;
+}
+
+bool SSet::checkSumOptimizedHS(DataType expected) {
+  int half_size = set_.size() / 2;
```
Comment (denislos) [inline subset/SSet.cpp:41]: I think it is better to use std::size_t here instead of int

File: subset/SSet.cpp [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  auto firstit = firstlist.begin();
+  auto secondit = secondlist.begin();
+
+  std::sort(firstlist.begin(), firstlist.end());
+  std::sort(secondlist.begin(), secondlist.end());
+
+  for (firstit = firstlist.begin(); firstit != firstlist.end(); firstit++) {
+    for (secondit = secondlist.begin(); secondit != secondlist.end(); secondit++) {
+      if ((*firstit == 0) || (*secondit == 0)) return true;
+      int sum = *firstit + *secondit;
```
Comment (denislos) [inline subset/SSet.cpp:57]: It should be DataType or auto in accordance with logic

File: subset/SSet.cpp [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+         checkSumOptimizedRecursive(i - 1, expected - set_[i]);
+}
+
+std::vector<SSet::DataType> SSet::subsetSums(std::vector<DataType>* set_p) {
+  std::vector<DataType> set = *set_p;
+  long total = 1 << set.size();  // total number of subsets = size of power set = 2^n
+  std::vector<SSet::DataType> sums(total, 0);
+  sums[1] = set[0];
+  int effectiveBits = 1, prevPowOf2 = 1;
+  for (long long i = 2; i < total; ++i) {
```
Comment (denislos) [inline subset/SSet.cpp:29]: There is a decorrelation with `long` and `long long` types used. I suggest to use std::size_t or std::uint64_t

Comment (denislos) [reply subset/SSet.cpp:29]: I think the same logic can be applied to `effectiveBits` and `prevPowOf2`

**Pull request merged**